shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -280,18 +280,25 @@ class ConstraintSolver extends NormalForms { self: Typer =>
           (if (cctx._1.headOption.exists(_ is lhs.prov)) cctx._1 else lhs :: cctx._1)
           ->
           (if (cctx._2.headOption.exists(_ is rhs.prov)) cctx._2 else rhs :: cctx._2)
-        else (lhs :: Nil) ->
-          (nestedProv.map(prov => rhs.withProv(prov) :: Nil).getOrElse(rhs :: Nil))
+        else (lhs :: Nil) -> {
+          println(s"Adding ${nestedProv.map(pv => pv.desc ++ pv.chain.toString())} nested prov for $lhs and $rhs")
+          val a = nestedProv.map(prov => rhs.withProv(prov) :: Nil).getOrElse(rhs :: Nil)
+          println(s"Added $a nested prov for $lhs and $rhs")
+          a
+        }
       )
     }
     def recImpl(lhs: SimpleType, rhs: SimpleType)
           (implicit raise: Raise, cctx: ConCtx): Unit =
     // trace(s"C $lhs <! $rhs") {
-    trace(s"C $lhs <! $rhs    (${cache.size})") {
+    trace(s"C $lhs <! $rhs    (${cache.size}) where lhs is ${lhs.getClass.getSimpleName} and rhs is ${rhs.getClass.getSimpleName} and cctx is $cctx}") {
     // trace(s"C $lhs <! $rhs  ${lhs.getClass.getSimpleName}  ${rhs.getClass.getSimpleName}") {
       // println(s"[[ ${cctx._1.map(_.prov).mkString(", ")}  <<  ${cctx._2.map(_.prov).mkString(", ")} ]]")
       // println(s"{{ ${cache.mkString(", ")} }}")
       
+      lazy val provChain = Some(NestedTypeProvenance(cctx._1 reverse_::: cctx._2))
+      lazy val revProvChain = Some(NestedTypeProvenance((cctx._1 reverse_::: cctx._2).reverse))
```

Comment:
I think this is equivalent (but more efficient!):
```suggestion
      lazy val revProvChain = Some(NestedTypeProvenance((cctx._2 reverse_::: cctx._1)))
```